kernel/syscalls/README.md

```diff
@@ -1,18 +1,58 @@
 # System calls
 
+System call commonly abbreviated to syscall makes an entry point to execute a specific user program's request to a service from the kernel. The operating system kernel runs in privileged mode to protect sensitive software or hardware parts from other software components. User application executing in unprivileged mode does not have access to protected data. Performing a hardware interrupt or conducting a trap handled by the kernel, the user application can obtain sensitive data from the kernel, e.g. information about all processes running in the operating system.
 
+## Declaration and definition
 
+In Phoenix-RTOS declarations and definitions of the system calls are located in a `libphoenix` library. The list of the all system calls is placed in `phoenix-rtos-kernel/include/syscalls.h` grouped by the categories.
 
+The system calls declaration should be placed in the appropriate header file in `libphoenix` library, referring to the syscall's category.
+The system calls definitions are placed in `arch/$(TARGET_SUFF)/syscalls.S` and are created based on syscall's list in preprocessing phase of the compilation. Each definition triggers an exception (e.g. SuperVisor Call - svc instruction for ARM Cortex-M or Cortex-A) with an appropriate syscall identification number handled by the kernel in privileged mode. Moreover, arguments to syscall are put on stack and redirected to handler.
 
+Handlers' definitions for system calls are located in `phoenix-rtos-kernel/syscalls.c`. Each handler should contain an appropriate return type consistent with the declaration in `libphoenix` and take stack pointer as a function argument. The syscall's parameters can be obtained from the user stack using the definition `GETFROMSTACK(stack ptr, arg type, var, id)`.
 
+Phoenix-RTOS in kernel mode has access to user memory maps so that pointer to data in user space can be passed as an argument to system call and used in kernel space.
 
+## Adding syscall
 
+An example of adding system call is conducted based on `threadsinfo` syscall. The procedure should be performed as follow:
 
+1. Add syscall's declaration in `libphoenix`, e.g. `libphoenix/inlude/sys/threads.h`:
 
+```C
+    extern int threadsinfo(int n, threadinfo_t *info);
+```
 
+2. Update syscall list in `phoenix-rtos-kernel/include/syscalls.h`:
 
+```C
+    ID(threadsinfo) \
```

Comment:
I would add a note about the vfork syscall or - even better - would advise that new syscalls should be added on the end of the list.